cranelift-filetests/src/function_runner.rs

```diff
@@ -1,53 +1,301 @@
 use cranelift_codegen::isa::TargetIsa;
 use cranelift_codegen::{settings, Context};
 use cranelift_native::builder as host_isa_builder;
+use cranelift_reader::Parser;
+use memmap::Mmap;
 use memmap::MmapMut;
 
 /// Run a function on a host
 pub struct FunctionRunner {
     function: Function,
     isa: Box<dyn TargetIsa>,
+    action: FunctionRunnerAction,
+}
+
+/// Build a function runner
+pub struct FunctionRunnerBuilder {
+    function: Function,
+    isa: Option<Box<dyn TargetIsa>>,
+    action: FunctionRunnerAction,
+}
+
+#[derive(Clone, Debug)]
+pub enum FunctionRunnerAction {
+    Print,
+    Test,
+    TestWithResult { operator: Operator, expected: Value },
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Operator {
+    EQ,
+    NEQ,
+}
+
+impl Display for Operator {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.write_str(match self {
+            Self::EQ => "==",
+            Self::NEQ => "!=",
+        })
+    }
+}
+
+impl FromStr for Operator {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "==" => Ok(Self::EQ),
+            "!=" => Ok(Self::NEQ),
+            _ => Err(format!("Unsupported operator {}", s)),
+        }
+    }
+}
+
+macro_rules! try_into_int {
+    ($type: ty, $name: ident) => {
+        struct $name {}
+        impl $name {
+            fn try_into(d: ConstantData) -> Result<$type, String> {
+                if d.len() == (mem::size_of::<$type>()) {
+                    let v = d.into_vec().into_iter().rev();
+                    let mut r: $type = 0;
+                    for b in v {
+                        let (mut shifted, _) = r.overflowing_shl(8);
+                        shifted |= b as $type;
+                        r = shifted;
+                    }
+                    Ok(r)
+                } else {
+                    Err(format!(
+                        "Incorrect vector size: {}, expected {}",
+                        d.len(),
+                        mem::size_of::<$type>()
+                    ))
+                }
+            }
+        }
+    };
+}
+
+try_into_int!(u8, ConstantDataU8);
+try_into_int!(u16, ConstantDataU16);
+try_into_int!(u32, ConstantDataU32);
+try_into_int!(u64, ConstantDataU64);
+try_into_int!(u128, ConstantDataU128);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Value {
+    Bool(bool),
+    F32(f32),
+    F64(f64),
+    I8(u8),
+    I16(u16),
+    I32(u32),
+    I64(u64),
+    I128(u128),
+}
+
+impl Value {
+    fn parse_as(s: &str, ty: types::Type) -> Result<Self, String> {
+        if ty.is_vector() {
+            return Err(format!("Expected a non-vector type, not {}", ty));
+        }
+
+        macro_rules! consume {
+            ( $ty:ident, $match_fn:expr ) => {{
+                ConstantData::default().append($match_fn)
+            }};
+        }
+
+        fn boolean_to_vec(value: bool) -> Vec<u8> {
```

Comment:
This function is used once, can we inline it at its single use, and then remove it instead?